protocol/superchainerc20/convert.md

```diff
@@ -0,0 +1,196 @@
+>- It makes the token leaner. Users will interact with `convert()` only a few times and in a context similar to deposit and withdraw.
+>- The `SuperchainERC20` representation for migrated liquidity will be as close as possible to newly deployed `SuperchainERC20`s.
+>- The `L2StandardBridge` needs to be involved because of burn and mint. Having the functionality in the token would have added additional external calls.
+
+
+### Access Control
+
+To check input addresses are valid, we will need to:
+
+- Check `SuperchainERC20` is valid: This depends on how the Factory is built.
+    - Store addresses: The simplest method is to store in the Factory a mapping that tracks the address of the corresponding `SuperchainERC20` for each L1 token (and other factors, such as decimals).
+    - Deterministic address: The Factory address is known, and so is the L1 token address, so the `SuperchainERC20` address can potentially be predicted. This heavily depends on the final choice of salt.
+- **Check `OptimismMintableERC20` is valid**: This involves verifying that the address was deployed from a valid factory. It is still undecided which Factories should be whitelisted. Possible approaches include:
+    - **Use a Registry**: An entity can update the registry at inception, and whitelisted factories can update it upon deployment. The Convert function will check mappings in this registry to allow a pair.
+    - **Use deterministic address**: This is only possible for some factories.
+        - The [new factory](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/OptimismMintableERC20Factory.sol) uses CREATE2 and will have a single representation. Deterministic address computation can verify the validity of the `OptimismMintableERC20`.
+        - The [old factory](https://github.com/ethereum-optimism/optimism-legacy/blob/develop/packages/contracts-bedrock/contracts/universal/OptimismMintableERC20Factory.sol) uses CREATE, meaning that the only possible approach is to iterate through the potential nonces. The maximum used nonce is not available in the contract, so the best approach would be to have some kind of registry (single-time update).
+- **Check `SuperchainERC20` address corresponds to the correct L1 address**: This is verified during the same validity check of the `SuperchainERC20`.
+
+A possible implementation would look like this:
+
+```solidity
+
+ISuperchainERC20Factory public superFactory = ISuperchainERC20Factory(factoryAddr);
+
+function isValidPair(address _legacyAddr, address _superAddr) internal view returns (bool) {
+  IOptimismMintableERC20 legacyToken = IOptimismMintableERC20(_legacyAddr);
+  address l1address = legacyToken.l1Token();
+  if (!registry.legacyDeployments(l1address)) {
+      return false;
+  }
+  return superFactory.deployment(l1address) == _superAddr;
+}
```

Comment:
This solidity does not check the condition:

> Check `OptimismMintableERC20` is valid

Any L2 contract could implement a `l1Token` function that can claim whatever here, right? To be unified, and merge its own supply with that of the real L2 token?

The `_legacyAddr` needs to be checked to be registered by the contract, or to be generated from a valid `OptimismMintableERC20` salt, to verify the deployment. Maybe ideally with a code-hash sanity check.

Reply:
We will want to figure out these details as part of the spec but if we go with the 1 time backport approach then we can get around needing to compute addresses on chain